promql/histogram_stats_iterator.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2024 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package promql
+
+import (
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/model/value"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+)
+
+type histogramStatsIterator struct {
+	chunkenc.Iterator
+
+	hReader *histogram.Histogram
+	lastH   *histogram.Histogram
+
+	fhReader *histogram.FloatHistogram
+	lastFH   *histogram.FloatHistogram
+}
+
+func NewHistogramStatsIterator(it chunkenc.Iterator) chunkenc.Iterator {
+	return &histogramStatsIterator{
+		Iterator: it,
+		hReader:  &histogram.Histogram{},
+		fhReader: &histogram.FloatHistogram{},
+	}
+}
+
+func (f *histogramStatsIterator) AtHistogram(h *histogram.Histogram) (int64, *histogram.Histogram) {
```

Comment:
I guess the doc comment here and below for `AtFloatHistogram` should mention that the rest detection only works if you don't switch between `AtHistogram` and `AtFloatHistogram`. (PromQL only ever uses the latter, so that should be fine in practice.)